Properly generate guards for any and unknown types

ritave opened the pull request:
fixes #182

Notice that this also generates proper types for union and intersection types as they are automatically collapsed by ts-morph

rhys-vdw commented:
Appreciate the PR, but could you please add test coverage @ritave?

rhys-vdw commented:
@ritave actually I think the problem is here:

https://github.com/rhys-vdw/ts-auto-guard/blob/559b3b9f53f0a24bdda1db2e0a927aa25a301281/src/index.ts#L936

Should be:

```ts
`return (\n${conditions || 'true'}\n)\n}\n`
```

Currently that is inlining `null` as in an absence of conditions. This is handled correctly in other functions where it simply won't check the properties (more desirable). For example we don't want this:

```ts
type Foo = any[]

// unnecessary
function isFoo(o: any): o is Foo {
  return Array.isArray(o) && o.every(() => true)
}

// good
function isFoo(o: any): o is Foo {
  return Array.isArray(o);
}
```

rhys-vdw commented:
Ah yes, that exact thing is going wrong in the test. Can you please add test cases for `type Foo = any[]` as well as the combinations you included in #182?

ritave commented:
@rhys-vdw I've added test cases, I've also changed the generation to allow primitive type aliases. Specifically `export type UnknownAndString = string & unknown` is equal to `export type UnknownAndString = string` and no type guard is generated for it. For completeness sake, I think it really should be generated.

rhys-vdw commented:
> Specifically export type UnknownAndString = string & unknown is equal to export type UnknownAndString = string and no type guard is generated for it.

@ritave I agree. This change was introduced here:
https://github.com/rhys-vdw/ts-auto-guard/pull/171#issuecomment-1014726266

I assume you're using the `exportAll` feature?

rhys-vdw commented:
Hm... I feel like @ritave might be sad if I merge this in without an option to disable generation of simple type aliases. @ritave I do agree that it's confusing (and therefore bad) behaviour to change how that works with and without the `exportAll` option so I'm tempted to just merge this. I don't think this would be a breaking change, just adds some extra guards that you wouldn't use. You should be able to remove them with tree shaking.

Let me know what you think before I merge.

ritave commented:
@rhys-vdw I prefer having guards generated for primitive types as well. We're using openrpc generated types and guards for those. Having guards, even for simple types ensures that changing the json itself won't force us to change the implementation since the guards would be re-generated properly.

rhys-vdw commented:
Sorry @ritave I mentioned you in my last comment instead of @kernwig (that must have been confusing). I just edited it.

> Since it's an opinionated issue, I'd be happy with a cli flag.

Yeah, well... Honestly I don't think it matters if you generate extra unused functions. I wonder if a better solution would be whitelisting e.g.

// generate types for _this_ file, instead of my-dependency
```ts
import type { Foo, Bar, Baz } from "my-dependency";

export { Foo, Bar } // exclude  unwanted Baz type
```

I'll give @kernwig a chance to chime in before I merge, but I'm strongly in behaviour of not having an option for excluding primitives in particular. IMO it's an arbitrary distinction which doesn't address the larger question of "how do I whitelist the functions I want to export" (which was my intention with the JSDoc style exporting).

kernwig commented:
@ritave I don't mind if the type guards for primitive types get generated, so long as they aren't used in place of more efficient checks in guards generated for interfaces. They will just tree-shake out. For union alias, that makes sense to have a guard and may be worth using within guards for interfaces; I don't think I considered or tested that use case.

What really got my attention to begin with was when there were two guards generated for the same primitive alias (like string), and the guard for an interface would call the wrong one. It was computationally the same, but misleading to my human eyes.

rhys-vdw commented:
> It was computationally the same, but misleading to my human eyes.

Ah true. Well if that's still the case then we should open an issue for it separately I guess.

> so long as they aren't used in place of more efficient checks in guards generated for interfaces.

I don't consider a single stack frame to be significant inefficiency in JS programs, but always open to improvements.

Thanks for the response @kernwig. I'm going to go ahead and merge this one. If there are any issues to your build you can hold off upgrading or open an issue to discuss the ideal behaviour.

rhys-vdw commented:
Thanks @ritave, I've released version 2.3.0 and invited you on as a collaborator. Feel free to contribute as much or as little as you please. If you accept and do want to make any changes, please open an issue for discussion and allow one collaborator to review your PRs before merging.

kernwig commented:
This did indeed "break" my fix. 

The change here adds these... useless to me but no objection:

```typescript
+export function isCognitoId(obj: any, _argumentName?: string): obj is CognitoId {
+  return typeof obj === "string";
+}
+
+export function isDeviceId(obj: any, _argumentName?: string): obj is DeviceId {
+  return typeof obj === "string";
+}
```

But then every single string comparison is again calling isCognitoId. Here's a example diff:

```typescript
 export function isRelationship(obj: any, _argumentName?: string): obj is Relationship {
   return (
     ((obj !== null && typeof obj === "object") || typeof obj === "function") &&
-    typeof obj.userId === "string" &&
-    typeof obj.deviceId === "string"
+    (isCognitoId(obj.userId) as boolean) &&
+    (isCognitoId(obj.deviceId) as boolean)
   );
 }
```

`isCognitoId(obj.deviceId)` is functional, but absolutely misleading. I don't think this in an opinionated issue. It isn't just string, every primitive type now calls the function for the first alias to it.

Please either call the correct guard (`isDeviceId`), or restore [my fix](https://github.com/rhys-vdw/ts-auto-guard/pull/171/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80) to use `typeof` for primitives within objects.
